docs/docs/dev_docs/contracts/portals/main.md

```diff
@@ -0,0 +1,331 @@
+
+```solidity title="IOutbox.sol"
+struct L2ToL1Msg {
+    DataStructures.L2Actor sender;
+    DataStructures.L1Actor recipient;
+    bytes32 content;
+}
+
+function consume(DataStructures.L2ToL1Msg memory _message) 
+    external 
+    returns (bytes32 entryKey);
+```
+
+As noted earlier, the portal contract should check that the sender is as expected. In the example below, we support only one sender contract (stored in `l2TokenAddress`) so we can just pass it as the sender, that way we will only be able to consume messages from that contract. If multiple senders are supported, you could use a have `mapping(address => bool) allowed` and check that `allowed[msg.sender]` is `true`.
+
+```solidity title="TokenPortal.sol"
+function withdraw(uint256 _amount, address _recipient, bool _withCaller)
+    external
+    returns (bytes32)
+  {
+    // Create the message structure
+    DataStructures.L2ToL1Msg memory message = DataStructures.L2ToL1Msg({
+      sender: DataStructures.L2Actor(l2TokenAddress, version),
+      recipient: DataStructures.L1Actor(address(this), block.chainid),
+      content: Hash.sha256ToField(
+        abi.encodeWithSignature(
+          "withdraw(uint256,address,address)",
+          _amount,
+          _recipient,
+          _withCaller ? msg.sender : address(0)
+        )
+        )
+    });
+
+    // Consume the message
+    bytes32 entryKey = registry.getOutbox().consume(message);
+
+    // Transfer the tokens to the user
+    underlying.transfer(_recipient, _amount);
+
+    return entryKey;
+  }
+```
+
+
+## How to deploy a contract with a portal
+- Deploy L1 
+- Deploy l2
+- Initialize l1 with l2 address.
+
+## Standards
+
+### Structure of messages
+The application developer should consider creating messages that follow a structure like function calls, e.g., using a function signature and arguments. This will make it easier to ensure that they are not making messages that could be misinterpreted by the recipient. 
+
+An example of a bad format would be using `amount, token_address, recipient_address` as the message for a withdraw function and `amount, token_address, on_behalf_of_address` for a deposit function. Any deposit could then also be mapped to a withdraw or vice versa. 
+
+```solidity 
+// Don't to this!
+bytes memory message = abi.encode(
+  _amount, 
+  _token, 
+  _to
+);
+
+// Do this!
+bytes memory message abi.encodeWithSignature(
+  "withdraw(uint256,address,address)", 
+  _amount, 
+  _token, 
+  _to
+);
+```
+
+### Error Handling
+
+Handling error when moving cross chain can quickly get tricky. Since the L1 and L2 calls are practically async and independent of each other, the L1 part of a deposit might execute just fine, with the L2 part failing. If not handling this well, the funds might be lost forever! The contract builder should therefore consider in what ways his application can fail cross chain, and handle those cases explicitly.
```

Comment:
```suggestion
Handling error when moving cross chain can quickly get tricky. Since the L1 and L2 calls are practically async and independent of each other, the L1 part of a deposit might execute just fine, with the L2 part failing. If this is not handled well, the funds may be lost forever! The contract builder should therefore consider ways their application can fail cross chain, and handle all cases explicitly.
```